src/AutomaticScheduling.h

```diff
@@ -0,0 +1,64 @@
+#ifndef HALIDE_AUTOMATIC_SCHEDULING_H
+#define HALIDE_AUTOMATIC_SCHEDULING_H
+
+/** \file
+ * Defines various automatic scheduling routines.
+ */
+
+#include "IR.h"
+
+namespace Halide {
+
+class Func;
+
+typedef enum {
+    ComputeRootAllStencils,
+    ParallelizeOuter,
+    VectorizeInner
+} AutoScheduleStrategy;
+
+namespace Internal {
+
+/** Base class for all automatic scheduling strategy implementations. */  
+class AutoScheduleStrategyImpl {
+public:
+    /** Apply the schedule strategy to the pipeline. 'func' should
```

Comment:
I know the inherited classes don't have any member variables yet, and the design is such that this is unlikely to be the case in the future. But it would probably still be a safe choice to add a virtual destructor, since we free these via the base class. It's really easy to forget adding this later.